### ISSUE TITLE
[META] Layout Followup Issues

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#345]
In https://github.com/backdrop/backdrop-issues/issues/86, we're working on completing the initial introduction of Layouts into Backdrop. That issue took us far, including introducing Layout module, simplifying Block module, providing contexts, access rules, and an expanded block API. However, there are many, many followups to be done. This issue serves as a listing of sub-issues, which may or may not yet have their own dedicated issues.

edit: I have created a GitHub "Project" for all the Layout related issues that were not completed as part of this meta. To view them all see: https://github.com/backdrop/backdrop-issues/projects/13

## Completed and merged

- [x] #372 | Add TouchPunch jQuery library to core to make our drag and drop work on touch devices.
- [x] #449 | Create a list of blocks and the modules that provide them
- [x] #570 | Automatically remove blocks from layouts when a module is uninstalled
- [x] #578 | [UX] Better layout conditions display (currently needs work)
- [x] #597 | Resolve conflicting style sheet names
- [x] #698 | Add additional layouts so sites can be built without needing to add either custom or contrib layouts.
- [x] #767 | Add hooks for Layout presave/insert/update
- [x] #791 | Add function to get current active layout on a page
- [x] #891 | Clean up the edit layout interface
- [x] #929 | Add a Views-like search box when adding new blocks
- [x] #943 | [UX] Canceling adding a condition should return to the "Add block" dialog (combined with #1020).
- [x] #952 | [UX] Move Layout Title Settings to a dialog
- [x] #958 | Update manager not configured for layouts
- [x] #998 | Clarify Layout terminology
- [x] #1017 | [UX] Block visibility conditions: Improve the is/is-not setting for "Front page" condition.
- [x] #1018 | Add a "Content type" visibility condition for blocks.
- [x] #1019 | [UX] Block settings: Consider renaming "Conditions" fieldset to "Visibility" or "Visibility conditions".
- [x] #1020 | [UX] Block UI: Dynamically load second-step visibility conditions configuration in the same dialog.
- [x] #1022 | [UX] Block configuration: "dirty" from flag fired even if user cancels adding a condition. (combined with #1020).
- [x] #1023 | Block conditions not displayed as a bullet list in the config block dialogue.
- [x] #1024 | Move the list of conditions in the block config dialogue below the "+ Add condition" link/button.
- [x] #1027 | backdrop_get_path() doesn't find layouts
- [x] #1028 | Make the layout list more useful and informative
- [x] #1076 | Add ability to add "existing node" to a layout
- [x] #1077 | Add a condition for selecting a specific entity ID
- [x] #1099 | Add a custom text block for non-reusable blocks (reusable blocks already are provided by Block module, but cannot be exported with the Layout config, as they are entities).
- [x] #1100 | Add blocks for adding individual fields from entities
- [x] #1300 | Add the ability to add a class to regions
- [x] #1547 | [UX] Move the list of visibility conditions in the layout add form below the "+ Add condition" link/button.
- [x] #1587 | Add Taxonomy term contexts
- [x] #1617 | Add ability to enable/disable certain layout templates (add a settings page for Layout module)
- [x] #1618 | Port the Panels "from block" title setting for layouts, allowing the title of a layout to be pulled from an individual block (combined with #952)
- [x] #1935 | Improve the aesthetic of the place-blocks page.

## Postponed/not doing
- [x] #1021 | [UX] Block config: Prevent the user from accidentally creating conditions that never/always display a block.

More issues are likely necessary. Let's expand this list as we find more suggestions for the module...

**Comment by quicksketch** [comment]
> Again in the create layout interface, if you type in the path before you select a Layout or a name for the Layout, when you're done writing the path, ajax runs and gives you an error about the name and field being required. Which is annoying. The validation should ideally be on final submit not on typing the path surely.

This is actually a much harder problem to solve than it appears. Layout module *requires* a name before it can start doing anything. The tempstore entry is saved by name, so we can't start doing anything until it is entered. Views has the same problem but gets around it by asking for the name of the view before you are dropped into the main views interface. I don't think introducing a new screen is a good idea for this. We could potentially hide the path option until the name is entered, but I feel like that's misleading the user into thinking the form is shorter than it really is, and would likely cause other UX problems.

> Should disable the "Create Layout" button till context ajax process is complete. 

> Layout 'content' panel does not populate with "main page content" block if I override an exisiting node path (eg node/50). I've tried, it doesnt happen. Node/% works. 

Both of these problems are being addressed in https://github.com/backdrop/backdrop-issues/issues/535.

**Comment by docwilmot** [comment]
>  Add a custom text block (with token support) for non-reusable blocks (reusable blocks already are provided by Block module, but cannot be exported with the Layout config, as they are entities).

Hadnt noticed this before. This is going to be confusing then, to have three different types of block? Here's a Big Bad Idea: get rid of the current "Custom Blocks" functionality and just have instances of one "Custom Block" block thats added through Layout builder's "Add block" UI.

**Comment by quicksketch** [comment]
@docwilmot Yeah that's actually pretty much exactly what I was thinking. Only we can't just get rid of all custom blocks in the 1.x cycle, Custom Blocks already a stand-alone module that you can easily disable.

**Comment by quicksketch** [comment]
> Add a custom text block (with token support) for non-reusable blocks (reusable blocks already are provided by Block module, but cannot be exported with the Layout config, as they are entities).

This has now changed. Blocks are in configuration in 1.3.0+ and are going to stick around. See https://github.com/backdrop/backdrop-issues/issues/1428 and https://github.com/backdrop/backdrop-issues/issues/1099.

**Comment by mikemccaffrey** [comment]
Does #1528 (Allow a layout to be selected on multiple paths) seem like something we can add to this meta list?

**Comment by docwilmot** [comment]
I've gone through the remaining issues and made a bunch of PRs. The remaining ones: #943, #1020 and #1022 are all related and awaiting some decision on how to proceed. I dont think I agree with #1025 either, so will wait for some further discussion. And I agree with @quicksketch #1021 is going to be very difficult to accomplish given how the `BlockAccess` class works.

#1300 should propbably use `Region Styles` but thats not fully implemented yet.

**Comment by docwilmot** [comment]
@klonos please test https://github.com/docwilmot/backdrop/tree/conditions. Fixes #943 and #1020. Still in-progress, was hoping to finish this morning, but out of time. And will add #1022 in final PR. Thanks.

**Comment by klonos** [comment]
I will do so, but most likely I'll be able to allocate time over this coming weekend. This one is + https://github.com/backdrop/backdrop/pull/1234 is on my list.

> ...And will add #1022 in final PR.

...while at it, can you tackle #1622 too that is related to the "dirty" flag not firing?

**Comment by docwilmot** [comment]
Now have  #943, #1020 and #1022 done and working in that PR, needs a nojs fallback so I can fix tests. Otherwise looks good.

**Comment by Graham-72** [comment]
@docwilmot Great progress! Thanks for all the effort you are putting into this.

**Comment by docwilmot** [comment]
Thx. These have been lingering too long. Just hope they get reviewed in time to get in 1.4.0.

**Comment by mikemccaffrey** [comment]
@docwilmot Sorry, I've been busy finding a place to rent in Portland, but will try to review things soon. Luckily, we still have a bit of time before the 1.4.0 release in May.

**Comment by docwilmot** [comment]
@mikemccaffrey no worries. 

And I'm done. The last remaining two I'm not sure if we should. Leaving them there.

**Comment by klonos** [comment]
...I need to find some time to go through the list of issues in the issue summary. Some of the issues I placed there are under the "Has PR" section while there's no PR for them.

**Comment by jenlampton** [comment]
Closing this issue since we now have the project https://github.com/backdrop/backdrop-issues/projects/13#card-21014976
